### DISCUSSION TITLE
실시간 라이브 플랫폼에서 발생하는 채팅 데이터는 어떻게 저장하고 처리할까요? Sql vs NoSql

### DISCUSSION THREAD

**Discussion by kangyuseok** [100-hours-a-week/discussions#30] [Deepdive]
1. 데이터베이스 선택
- 채팅 데이터는 구조화 되지 않은 메시지(사용자 메타데이터, 타임스탬프 등)
- 유연한 데이터 모델 vs 채팅 기록의 트랜잭션 무결성
2. 데이터 저장 및 인덱싱
- 시간 순서대로 정렬
- 특정 사용자의 메시지
- 특정 대화방의 대화
- 첨부파일 저장(이미지, 비디오, 파일)
3. 데이터 압축 및 보관
- 오래된 채팅 압축
- 일정 기간 지난 채팅
4. 실시간 처리 및 성능 최적화
- 실시간 통신 프로토콜 과 DB 연결 최적화
- 캐싱
- 사용자가 늘어남에 따라 분산 데이터베이스 시스템 또는 샤딩 고려
5. 보안 및 프라이버시
- 데이터 암호화
- 접근 제어(사용자가 자신이 속한 대화방의 메시지만 접근)
- 사용자 프라이버시(개인정보 보호법을 준수하여 사용자 데이터 처리 방식 고려)
6. 데이터 일관성 및 복구
- 백업 및 복구(데이터 손실 대비)
7. 확장성 및 유지보수
- 시스템의 성능 및 오류 모니터링
- 스키마의 빈번한 변경
- 사용자가 늘어났을 때 현재 DB구조가 이를 수용할 수 있는가?

**Comment by startupcodekr** [comment]
본인 프로젝트의 문제점 시나리오의 내용이 필요
